### PR TITLE
SONARJAVA-958 Migrate Unused private method to semantic analysis

### PR DESCRIPTION
**Pull request by merciesa**

### CODE REVIEW

File: java-checks/src/test/java/org/sonar/java/checks/targets/UnusedPrivateMethod.java [hunk 1]
```diff
@@ -51,26 +51,41 @@ private void readObjectNoData() throws java.io.ObjectStreamException {
...
+  private void variadic(String first, Object... objects) {
+  }
+
+  private void variadic(String... strings) { // Noncompliant {{Private method 'variadic' is never used.}}
+    // this is a false positive
+  }
+
+  public void usage() {
+    variadic("", new Object(), new Object());
+    variadic("", "", ""); // should resolve to 'String...' and not 'String, Object...'
```
Comment (benzonico) [inline java-checks/src/test/java/org/sonar/java/checks/targets/UnusedPrivateMethod.java]: This should lead to a creation of the corresponding jira ticket to fix semantic analysis.

Comment (benzonico) [inline java-checks/src/test/java/org/sonar/java/checks/targets/UnusedPrivateMethod.java]: Moreover I think this test should also include some generic methods.

File: java-checks/src/main/java/org/sonar/java/checks/UnusedPrivateMethodCheck.java [hunk 2]
```diff
@@ -46,47 +49,69 @@
...
+      memberTree.accept(new MethodVisitor(methodInvocationVisitor.calledPrivateMethods));
+    }
+  }
+
+  private static class MethodInvocationVisitor extends BaseTreeVisitor {
+    final Set<Symbol> calledPrivateMethods = new HashSet<>();
+
+    @Override
+    public void visitNewClass(NewClassTree tree) {
+      Symbol constructorSymbol = ((NewClassTreeImpl) tree).getConstructorIdentifier().symbol();
```
Comment (merciesa) [inline java-checks/src/main/java/org/sonar/java/checks/UnusedPrivateMethodCheck.java]: @benzonico your opinion on this? I am tempted to say we should expose the symbol of the constructor through the interface. NewClassTree.constructor()?

Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/UnusedPrivateMethodCheck.java]: Sounds correct.

File: java-checks/src/main/java/org/sonar/java/checks/UnusedPrivateMethodCheck.java [hunk 3]
```diff
@@ -19,24 +19,29 @@
...
-import org.sonar.java.bytecode.asm.AsmClass;
-import org.sonar.java.bytecode.asm.AsmMethod;
-import org.sonar.java.bytecode.visitor.BytecodeVisitor;
-import org.sonar.java.signature.MethodSignatureScanner;
-import org.sonar.java.signature.Parameter;
+import org.sonar.java.checks.methods.MethodInvocationMatcherCollection;
+import org.sonar.java.checks.methods.MethodMatcher;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ClassTree;
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/UnusedPrivateMethodCheck.java]: MINOR Remove this unused import 'org.sonar.plugins.java.api.tree.ClassTree'. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Pull request closed without merging**